File: buildingForest.rb
# Problem Link: http://www.codeforces.com/problemset/problem/195/E */http://www.codeforces.com/problemset/problem/195/E
MOD = 1e9+7

class DisjointSet
	def initialize(n)
		@parent = (0..n).to_a
		@depth = [0]*(n+1)
	end

	def get(x)
		return x if @parent[x] == x
		pa = @parent[x]
		@parent[x] = get(@parent[x])
		@depth[x] += @depth[pa]
		@parent[x]
	end

	def add(root, x, w)
		@parent[root] = x
		@depth[root] = w
	end

	def depth(n)
		get(n)
		@depth[n]
	end
end


n = gets.to_i
ds = DisjointSet.new(n)
ans = 0
ar = []
n.times do |idx|
	ar = gets.split.map(&:to_i)
	(1..ar[0]).each do |i|
		v = ar[2*i -1]
		x = ar[2*i]
		root = ds.get(v)
		ans = (ans + ds.depth(v) + x)%MOD
		ds.add(root, idx+1, ds.depth(v) +x)%MOD
	end
end
puts (ans%MOD).to_i

File: buildingForest.cpp
/* Problem Link: http://www.codeforces.com/problemset/problem/195/E */
#include <iostream>
#include <cstdio>
#include <vector>
using namespace std;
const int MAX_N = 1e5+7;
const int mod = 1e9+7;
int parent[MAX_N];
long long depth[MAX_N];
pair<int, long long> getp(int x)
{
	if (parent[x] == x) return make_pair(parent[x], depth[x]);
	pair<int, long long> ret = getp(parent[x]);
	parent[x] = ret.first;
	depth[x] = (depth[x] + ret.second)%mod;
	return make_pair(parent[x], depth[x]);
}
void add(int root, int n, long long x)
{
	parent[root] = n;
	depth[root] = x%mod;
}
int main()
{
	int n;
	cin>>n;
	long long ans = 0;
	for (int i = 1; i<= n; i++) parent[i] = i;
	for (int i = 1; i <= n; i++) {
		int k, v;
		long long x;
		cin>>k;
		for (int j = 1; j <= k; j++) {
			cin>>v>>x;
			int root = getp(v).first;
			ans = (ans + depth[v] + x)%mod;
			add(root, i, x + depth[v]);
		}
	}
	cout<<(ans+mod)%mod<<endl;
	return 0;
}